cmd/snapd-env-generator/main.c

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include<stdlib.h>
+#include<string.h>
+#include<stdio.h>
+
+#include "config.h"
+
+// Systemd environment generators work since version 233 which ships
+// in Ubuntu 17.10+
+int main(int argc, char **argv)
+{
+        const char *snap_bin_dir = SNAP_MOUNT_DIR"/bin";
+
+	char *path = getenv("PATH");
+	if (strstr(path, snap_bin_dir) != NULL) {
```

Comment:
Not a big deal but we could use strchr to iterate over PATH components and then use strcmp to ensure that we have a perfect fit. The current approach with `strstr` would match PATH that is `/snap/bin/baz:/usr/bin` for example.

Reply:
So how about that shell script? :smile: If I'm not mistaken, the whole thing could be:
```
#!/bin/sh
snap_bin_path="@SNAP_MOUNT_DIR@/bin"
if [ -n "${PATH##*${snap_bin_path}}" -a -n "${PATH##*${snap_bin_path}:*}" ]; then
    echo PATH=$PATH:${snap_bin_path}
fi
```